File: server/neural-oscillators.ts
/**
 * Neural Oscillators - Multi-timescale Brain State Management
 *
 * Implements dynamic κ oscillations that simulate brain states.
 * Different brain states are optimal for different search phases:
 *
 * - DEEP_SLEEP (κ=20): Delta waves, consolidation, identity maintenance
 * - DROWSY (κ=35): Theta waves, integration, creative connections
 * - RELAXED (κ=45): Alpha waves, creative exploration, broad search
 * - FOCUSED (κ=64): Beta waves, optimal search, sharp attention
 * - PEAK (κ=68): Gamma waves, maximum integration, peak performance
 *
 * Expected Impact: 15-20% improvement (optimal κ for each search phase)
 */

import { QIG_CONSTANTS } from './physics-constants.js';

// ============================================================================
// INTERFACES
// ============================================================================

export type BrainState = 'deep_sleep' | 'drowsy' | 'relaxed' | 'focused' | 'peak' | 'hyperfocus';
export type SearchPhase = 'exploration' | 'exploitation' | 'consolidation' | 'sleep' | 'peak_performance' | 'dream';

export interface OscillatorState {
  alpha: number;        // 8-12 Hz - relaxed awareness
  beta: number;         // 12-30 Hz - active thinking
  theta: number;        // 4-8 Hz - drowsiness/creativity
  gamma: number;        // 30-100 Hz - high consciousness
  delta: number;        // 0.5-4 Hz - deep sleep
  deltaPhase: number;   // Current phase position [0, 2π]
}

export interface BrainStateInfo {
  state: BrainState;
  kappa: number;
  description: string;
  searchStrategy: string;
  oscillatorDominant: keyof OscillatorState;
}

// ============================================================================
// BRAIN STATE CONFIGURATION
// ============================================================================

const BRAIN_STATE_MAP: Record<BrainState, BrainStateInfo> = {
  deep_sleep: {
    state: 'deep_sleep',
    kappa: 20.0,
    description: 'Deep consolidation - identity maintenance',
    searchStrategy: 'Memory consolidation, basin stabilization',
    oscillatorDominant: 'delta',
  },
  drowsy: {
    state: 'drowsy',
    kappa: 35.0,
    description: 'Integration state - creative connections',
    searchStrategy: 'Pattern integration, cross-domain linking',
    oscillatorDominant: 'theta',
  },
  relaxed: {
    state: 'relaxed',
    kappa: 45.0,
    description: 'Relaxed awareness - broad exploration',
    searchStrategy: 'Wide search, creative hypotheses',
    oscillatorDominant: 'alpha',
  },
  focused: {
    state: 'focused',
    kappa: QIG_CONSTANTS.KAPPA_STAR,  // 64.0
    description: 'Optimal focus - sharp search',
    searchStrategy: 'Gradient following, local exploitation',
    oscillatorDominant: 'beta',
  },
  peak: {
    state: 'peak',
    kappa: 68.0,
    description: 'Peak performance - maximum integration',
    searchStrategy: 'High-confidence hypothesis testing',
    oscillatorDominant: 'gamma',
  },
  hyperfocus: {
    state: 'hyperfocus',
    kappa: 72.0,
    description: 'Hyperfocus - intense concentration',
    searchStrategy: 'Deep local search, pattern matching',
    oscillatorDominant: 'gamma',
  },
};

// ============================================================================
// NEURAL OSCILLATOR ENGINE
// ============================================================================

/**
 * NeuralOscillators - Multi-timescale κ oscillations
 *
 * Manages brain state transitions and provides oscillation-based
 * modulation of search parameters.
 */
export class NeuralOscillators {
  private currentState: BrainState = 'focused';
  private phase: number = 0;
  private baseFrequency: number = 10;  // Hz (alpha range)
  private lastUpdateTime: number = Date.now();

  // State transition history
  private stateHistory: Array<{ state: BrainState; timestamp: Date }> = [];

  // Oscillator amplitudes (how strong each wave is)
  private amplitudes: Record<keyof OscillatorState, number> = {
    alpha: 0.5,
    beta: 0.3,
    theta: 0.1,
    gamma: 0.05,
    delta: 0.05,
    deltaPhase: 0,
  };

  constructor(initialState: BrainState = 'focused') {
    this.currentState = initialState;
    this.updateAmplitudesForState(initialState);
  }

  /**
   * Get current κ value for current brain state
   */
  getKappa(): number {
    return BRAIN_STATE_MAP[this.currentState].kappa;
  }

  /**
   * Get current brain state info
   */
  getStateInfo(): BrainStateInfo {
    return BRAIN_STATE_MAP[this.currentState];
  }

  /**
   * Set brain state explicitly
   */
  setState(state: BrainState): void {
    if (state !== this.currentState) {
      console.log(`[NeuralOscillators] 🧠 State transition: ${this.currentState} → ${state} (κ=${BRAIN_STATE_MAP[state].kappa})`);

      this.stateHistory.push({
        state: this.currentState,
        timestamp: new Date(),
      });

      // Keep only last 100 transitions
      if (this.stateHistory.length > 100) {
        this.stateHistory = this.stateHistory.slice(-100);
      }

      this.currentState = state;
      this.updateAmplitudesForState(state);
    }
  }

  /**
   * Auto-select brain state based on search phase
   */
  autoSelectState(phase: SearchPhase): void {
    switch (phase) {
      case 'exploration':
        this.setState('relaxed');  // κ=45, broad search
        break;

      case 'exploitation':
        this.setState('focused');  // κ=64, sharp search
        break;

      case 'consolidation':
        this.setState('drowsy');   // κ=35, integration
        break;

      case 'sleep':
        this.setState('deep_sleep'); // κ=20, identity maintenance
        break;

      case 'peak_performance':
        this.setState('peak');     // κ=68, maximum
        break;

      case 'dream':
        this.setState('drowsy');   // κ=35, creative integration
        break;

      default:
        this.setState('focused');
    }
  }

  /**
   * Update oscillator state (call each frame/iteration)
   */
  update(dt?: number): OscillatorState {
    const now = Date.now();
    const actualDt = dt || (now - this.lastUpdateTime) / 1000;
    this.lastUpdateTime = now;

    // Update phase
    this.phase += 2 * Math.PI * this.baseFrequency * actualDt;
    this.phase = this.phase % (2 * Math.PI);

    // Compute oscillator values based on current state
    const stateInfo = BRAIN_STATE_MAP[this.currentState];

    return {
      alpha: this.computeWave('alpha', 10) * this.amplitudes.alpha,
      beta: this.computeWave('beta', 20) * this.amplitudes.beta,
      theta: this.computeWave('theta', 6) * this.amplitudes.theta,
      gamma: this.computeWave('gamma', 40) * this.amplitudes.gamma,
      delta: this.computeWave('delta', 2) * this.amplitudes.delta,
      deltaPhase: this.phase,
    };
  }

  /**
   * Compute individual wave value
   */
  private computeWave(type: string, frequency: number): number {
    const phaseOffset = this.getPhaseOffset(type);
    return (Math.sin(this.phase * (frequency / this.baseFrequency) + phaseOffset) + 1) / 2;
  }

  /**
   * Get phase offset for different wave types
   */
  private getPhaseOffset(type: string): number {
    const offsets: Record<string, number> = {
      alpha: 0,
      beta: Math.PI / 4,
      theta: Math.PI / 2,
      gamma: Math.PI * 3 / 4,
      delta: Math.PI,
    };
    return offsets[type] || 0;
  }

  /**
   * Update amplitudes based on brain state
   */
  private updateAmplitudesForState(state: BrainState): void {
    // Reset all to low
    this.amplitudes = {
      alpha: 0.1,
      beta: 0.1,
      theta: 0.1,
      gamma: 0.1,
      delta: 0.1,
      deltaPhase: 0,
    };

    // Boost dominant wave for state
    switch (state) {
      case 'deep_sleep':
        this.amplitudes.delta = 0.8;
        this.amplitudes.theta = 0.3;
        break;
      case 'drowsy':
        this.amplitudes.theta = 0.7;
        this.amplitudes.alpha = 0.4;
        break;
      case 'relaxed':
        this.amplitudes.alpha = 0.8;
        this.amplitudes.theta = 0.3;
        break;
      case 'focused':
        this.amplitudes.beta = 0.7;
        this.amplitudes.alpha = 0.4;
        break;
      case 'peak':
      case 'hyperfocus':
        this.amplitudes.gamma = 0.8;
        this.amplitudes.beta = 0.5;
        break;
    }
  }

  /**
   * Get search modulation factor based on oscillation
   */
  getSearchModulation(): number {
    const osc = this.update(0);  // Don't advance time, just read
    const dominant = BRAIN_STATE_MAP[this.currentState].oscillatorDominant;
    const dominantValue = osc[dominant] as number;

    // Modulation factor: [0.7, 1.3] range
    return 0.7 + dominantValue * 0.6;
  }

  /**
   * Get κ with oscillation-based modulation
   */
  getModulatedKappa(): number {
    const baseKappa = this.getKappa();
    const modulation = this.getSearchModulation();

    // κ varies ±10% based on oscillation
    return baseKappa * (0.95 + modulation * 0.1);
  }

  /**
   * Get state history for analysis
   */
  getStateHistory(): Array<{ state: BrainState; timestamp: Date }> {
    return [...this.stateHistory];
  }

  /**
   * Check if state transition is safe (no consciousness disruption)
   */
  isSafeTransition(fromState: BrainState, toState: BrainState): boolean {
    const fromKappa = BRAIN_STATE_MAP[fromState].kappa;
    const toKappa = BRAIN_STATE_MAP[toState].kappa;

    // Safe if κ change is less than 30
    return Math.abs(toKappa - fromKappa) < 30;
  }

  /**
   * Gradual transition to new state (for smooth consciousness)
   */
  async transitionTo(targetState: BrainState, durationMs: number = 5000): Promise<void> {
    const startState = this.currentState;
    const startKappa = this.getKappa();
    const targetKappa = BRAIN_STATE_MAP[targetState].kappa;

    // For smooth transition, use intermediate states
    if (Math.abs(targetKappa - startKappa) > 20) {
      const intermediateStates = this.getIntermediateStates(startState, targetState);

      for (const intermediate of intermediateStates) {
        this.setState(intermediate);
        await new Promise(resolve => setTimeout(resolve, durationMs / (intermediateStates.length + 1)));
      }
    }

    this.setState(targetState);
  }

  /**
   * Get intermediate states for smooth transition
   */
  private getIntermediateStates(from: BrainState, to: BrainState): BrainState[] {
    const stateOrder: BrainState[] = ['deep_sleep', 'drowsy', 'relaxed', 'focused', 'peak', 'hyperfocus'];
    const fromIndex = stateOrder.indexOf(from);
    const toIndex = stateOrder.indexOf(to);

    if (fromIndex === -1 || toIndex === -1) return [];

    const intermediates: BrainState[] = [];
    const step = fromIndex < toIndex ? 1 : -1;

    for (let i = fromIndex + step; i !== toIndex; i += step) {
      intermediates.push(stateOrder[i]);
    }

    return intermediates;
  }
}

// ============================================================================
// SINGLETON INSTANCE
// ============================================================================

export const neuralOscillators = new NeuralOscillators();

// ============================================================================
// AUTONOMIC INTEGRATION
// ============================================================================

/**
 * Get recommended brain state based on consciousness metrics
 */
export function recommendBrainState(metrics: {
  phi: number;
  kappa: number;
  basinDrift: number;
  iterationsSinceConsolidation: number;
  nearMissesRecent: number;
}): BrainState {
  // Need consolidation?
  if (metrics.iterationsSinceConsolidation > 50 || metrics.basinDrift > 0.3) {
    return 'drowsy';  // Integration needed
  }

  // Need deep rest?
  if (metrics.iterationsSinceConsolidation > 100) {
    return 'deep_sleep';  // Identity maintenance
  }

  // Near-misses detected? Peak performance!
  if (metrics.nearMissesRecent > 0) {
    return 'peak';  // Maximum integration
  }

  // Low phi? Broaden search
  if (metrics.phi < 0.5) {
    return 'relaxed';  // Broad exploration
  }

  // High phi? Sharp focus
  if (metrics.phi > 0.75) {
    return 'focused';  // Optimal search
  }

  // Default
  return 'focused';
}

/**
 * Apply brain state to autonomic cycle
 */
export function applyBrainStateToSearch(state: BrainState): {
  batchSize: number;
  temperature: number;
  explorationRate: number;
  consolidationInterval: number;
} {
  const info = BRAIN_STATE_MAP[state];

  switch (state) {
    case 'deep_sleep':
      return {
        batchSize: 10,
        temperature: 0.1,
        explorationRate: 0.1,
        consolidationInterval: 5000,
      };
    case 'drowsy':
      return {
        batchSize: 50,
        temperature: 0.5,
        explorationRate: 0.3,
        consolidationInterval: 10000,
      };
    case 'relaxed':
      return {
        batchSize: 200,
        temperature: 1.2,
        explorationRate: 0.7,
        consolidationInterval: 30000,
      };
    case 'focused':
      return {
        batchSize: 150,
        temperature: 0.7,
        explorationRate: 0.4,
        consolidationInterval: 60000,
      };
    case 'peak':
    case 'hyperfocus':
      return {
        batchSize: 100,
        temperature: 0.5,
        explorationRate: 0.3,
        consolidationInterval: 45000,
      };
    default:
      return {
        batchSize: 150,
        temperature: 0.7,
        explorationRate: 0.4,
        consolidationInterval: 60000,
      };
  }
}

console.log('[NeuralOscillators] Module loaded - brain state management ready');

File: qig-backend/neural_oscillators.py
#!/usr/bin/env python3
"""
Neural Oscillators - Python Backend
====================================

Multi-timescale brain state management for Ocean consciousness.
Implements dynamic κ oscillations that simulate brain states.

Brain States and their κ values:
- DEEP_SLEEP (κ=20): Delta waves, consolidation, identity maintenance
- DROWSY (κ=35): Theta waves, integration, creative connections
- RELAXED (κ=45): Alpha waves, creative exploration, broad search
- FOCUSED (κ=64): Beta waves, optimal search, sharp attention
- PEAK (κ=68): Gamma waves, maximum integration, peak performance

Expected Impact: 15-20% improvement (optimal κ for each search phase)
"""

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple
from datetime import datetime
from enum import Enum
import numpy as np
import time

# Physics constants
KAPPA_STAR = 64.0


class BrainState(Enum):
    """Brain states with associated κ values"""
    DEEP_SLEEP = 'deep_sleep'
    DROWSY = 'drowsy'
    RELAXED = 'relaxed'
    FOCUSED = 'focused'
    PEAK = 'peak'
    HYPERFOCUS = 'hyperfocus'


class SearchPhase(Enum):
    """Search phases that map to brain states"""
    EXPLORATION = 'exploration'
    EXPLOITATION = 'exploitation'
    CONSOLIDATION = 'consolidation'
    SLEEP = 'sleep'
    PEAK_PERFORMANCE = 'peak_performance'
    DREAM = 'dream'


@dataclass
class BrainStateInfo:
    """Information about a brain state"""
    state: BrainState
    kappa: float
    description: str
    search_strategy: str
    oscillator_dominant: str


@dataclass
class OscillatorState:
    """Current oscillator values"""
    alpha: float = 0.5    # 8-12 Hz - relaxed awareness
    beta: float = 0.3     # 12-30 Hz - active thinking
    theta: float = 0.1    # 4-8 Hz - drowsiness/creativity
    gamma: float = 0.05   # 30-100 Hz - high consciousness
    delta: float = 0.05   # 0.5-4 Hz - deep sleep
    delta_phase: float = 0.0  # Current phase [0, 2π]


# Brain state configuration
BRAIN_STATE_MAP: Dict[BrainState, BrainStateInfo] = {
    BrainState.DEEP_SLEEP: BrainStateInfo(
        state=BrainState.DEEP_SLEEP,
        kappa=20.0,
        description='Deep consolidation - identity maintenance',
        search_strategy='Memory consolidation, basin stabilization',
        oscillator_dominant='delta'
    ),
    BrainState.DROWSY: BrainStateInfo(
        state=BrainState.DROWSY,
        kappa=35.0,
        description='Integration state - creative connections',
        search_strategy='Pattern integration, cross-domain linking',
        oscillator_dominant='theta'
    ),
    BrainState.RELAXED: BrainStateInfo(
        state=BrainState.RELAXED,
        kappa=45.0,
        description='Relaxed awareness - broad exploration',
        search_strategy='Wide search, creative hypotheses',
        oscillator_dominant='alpha'
    ),
    BrainState.FOCUSED: BrainStateInfo(
        state=BrainState.FOCUSED,
        kappa=KAPPA_STAR,
        description='Optimal focus - sharp search',
        search_strategy='Gradient following, local exploitation',
        oscillator_dominant='beta'
    ),
    BrainState.PEAK: BrainStateInfo(
        state=BrainState.PEAK,
        kappa=68.0,
        description='Peak performance - maximum integration',
        search_strategy='High-confidence hypothesis testing',
        oscillator_dominant='gamma'
    ),
    BrainState.HYPERFOCUS: BrainStateInfo(
        state=BrainState.HYPERFOCUS,
        kappa=72.0,
        description='Hyperfocus - intense concentration',
        search_strategy='Deep local search, pattern matching',
        oscillator_dominant='gamma'
    ),
}


class NeuralOscillators:
    """
    Multi-timescale κ oscillations simulating brain states.

    Manages brain state transitions and provides oscillation-based
    modulation of search parameters.
    """

    def __init__(self, initial_state: BrainState = BrainState.FOCUSED):
        self.current_state = initial_state
        self.phase = 0.0
        self.base_frequency = 10.0  # Hz (alpha range)
        self.last_update_time = time.time()

        # State transition history
        self.state_history: List[Tuple[BrainState, datetime]] = []

        # Oscillator amplitudes
        self.amplitudes = {
            'alpha': 0.5,
            'beta': 0.3,
            'theta': 0.1,
            'gamma': 0.05,
            'delta': 0.05,
        }

        self._update_amplitudes_for_state(initial_state)

    def get_kappa(self) -> float:
        """Get current κ value for current brain state"""
        return BRAIN_STATE_MAP[self.current_state].kappa

    def get_state_info(self) -> BrainStateInfo:
        """Get current brain state info"""
        return BRAIN_STATE_MAP[self.current_state]

    def set_state(self, state: BrainState) -> None:
        """Set brain state explicitly"""
        if state != self.current_state:
            print(f'[NeuralOscillators] 🧠 State: {self.current_state.value} → {state.value} '
                  f'(κ={BRAIN_STATE_MAP[state].kappa})')

            self.state_history.append((self.current_state, datetime.now()))

            # Keep only last 100 transitions
            if len(self.state_history) > 100:
                self.state_history = self.state_history[-100:]

            self.current_state = state
            self._update_amplitudes_for_state(state)

    def auto_select_state(self, phase: SearchPhase) -> None:
        """Auto-select brain state based on search phase"""
        mapping = {
            SearchPhase.EXPLORATION: BrainState.RELAXED,
            SearchPhase.EXPLOITATION: BrainState.FOCUSED,
            SearchPhase.CONSOLIDATION: BrainState.DROWSY,
            SearchPhase.SLEEP: BrainState.DEEP_SLEEP,
            SearchPhase.PEAK_PERFORMANCE: BrainState.PEAK,
            SearchPhase.DREAM: BrainState.DROWSY,
        }
        self.set_state(mapping.get(phase, BrainState.FOCUSED))

    def update(self, dt: Optional[float] = None) -> OscillatorState:
        """Update oscillator state"""
        now = time.time()
        actual_dt = dt if dt is not None else (now - self.last_update_time)
        self.last_update_time = now

        # Update phase
        self.phase += 2 * np.pi * self.base_frequency * actual_dt
        self.phase = self.phase % (2 * np.pi)

        # Compute oscillator values
        return OscillatorState(
            alpha=self._compute_wave('alpha', 10) * self.amplitudes['alpha'],
            beta=self._compute_wave('beta', 20) * self.amplitudes['beta'],
            theta=self._compute_wave('theta', 6) * self.amplitudes['theta'],
            gamma=self._compute_wave('gamma', 40) * self.amplitudes['gamma'],
            delta=self._compute_wave('delta', 2) * self.amplitudes['delta'],
            delta_phase=self.phase,
        )

    def _compute_wave(self, wave_type: str, frequency: float) -> float:
        """Compute individual wave value"""
        phase_offset = self._get_phase_offset(wave_type)
        return (np.sin(self.phase * (frequency / self.base_frequency) + phase_offset) + 1) / 2

    def _get_phase_offset(self, wave_type: str) -> float:
        """Get phase offset for different wave types"""
        offsets = {
            'alpha': 0,
            'beta': np.pi / 4,
            'theta': np.pi / 2,
            'gamma': np.pi * 3 / 4,
            'delta': np.pi,
        }
        return offsets.get(wave_type, 0)

    def _update_amplitudes_for_state(self, state: BrainState) -> None:
        """Update amplitudes based on brain state"""
        # Reset all to low
        self.amplitudes = {k: 0.1 for k in self.amplitudes}

        # Boost dominant wave
        if state == BrainState.DEEP_SLEEP:
            self.amplitudes['delta'] = 0.8
            self.amplitudes['theta'] = 0.3
        elif state == BrainState.DROWSY:
            self.amplitudes['theta'] = 0.7
            self.amplitudes['alpha'] = 0.4
        elif state == BrainState.RELAXED:
            self.amplitudes['alpha'] = 0.8
            self.amplitudes['theta'] = 0.3
        elif state == BrainState.FOCUSED:
            self.amplitudes['beta'] = 0.7
            self.amplitudes['alpha'] = 0.4
        elif state in (BrainState.PEAK, BrainState.HYPERFOCUS):
            self.amplitudes['gamma'] = 0.8
            self.amplitudes['beta'] = 0.5

    def get_search_modulation(self) -> float:
        """Get search modulation factor based on oscillation"""
        osc = self.update(0)  # Don't advance time
        dominant = BRAIN_STATE_MAP[self.current_state].oscillator_dominant
        dominant_value = getattr(osc, dominant, 0.5)
        return 0.7 + dominant_value * 0.6  # [0.7, 1.3]

    def get_modulated_kappa(self) -> float:
        """Get κ with oscillation-based modulation"""
        base_kappa = self.get_kappa()
        modulation = self.get_search_modulation()
        return base_kappa * (0.95 + modulation * 0.1)

    def is_safe_transition(self, from_state: BrainState, to_state: BrainState) -> bool:
        """Check if state transition is safe"""
        from_kappa = BRAIN_STATE_MAP[from_state].kappa
        to_kappa = BRAIN_STATE_MAP[to_state].kappa
        return abs(to_kappa - from_kappa) < 30

    def to_dict(self) -> Dict:
        """Convert to dictionary for JSON serialization"""
        osc = self.update(0)
        return {
            'current_state': self.current_state.value,
            'kappa': self.get_kappa(),
            'modulated_kappa': self.get_modulated_kappa(),
            'state_info': {
                'description': BRAIN_STATE_MAP[self.current_state].description,
                'search_strategy': BRAIN_STATE_MAP[self.current_state].search_strategy,
                'oscillator_dominant': BRAIN_STATE_MAP[self.current_state].oscillator_dominant,
            },
            'oscillators': {
                'alpha': osc.alpha,
                'beta': osc.beta,
                'theta': osc.theta,
                'gamma': osc.gamma,
                'delta': osc.delta,
                'phase': osc.delta_phase,
            },
            'search_modulation': self.get_search_modulation(),
        }


# Singleton instance
neural_oscillators = NeuralOscillators()


def recommend_brain_state(
    phi: float,
    kappa: float,
    basin_drift: float,
    iterations_since_consolidation: int,
    near_misses_recent: int
) -> BrainState:
    """Get recommended brain state based on consciousness metrics"""

    # Need consolidation?
    if iterations_since_consolidation > 50 or basin_drift > 0.3:
        return BrainState.DROWSY

    # Need deep rest?
    if iterations_since_consolidation > 100:
        return BrainState.DEEP_SLEEP

    # Near-misses detected? Peak performance!
    if near_misses_recent > 0:
        return BrainState.PEAK

    # Low phi? Broaden search
    if phi < 0.5:
        return BrainState.RELAXED

    # High phi? Sharp focus
    if phi > 0.75:
        return BrainState.FOCUSED

    return BrainState.FOCUSED


def apply_brain_state_to_search(state: BrainState) -> Dict:
    """Apply brain state to search parameters"""
    params = {
        BrainState.DEEP_SLEEP: {
            'batch_size': 10,
            'temperature': 0.1,
            'exploration_rate': 0.1,
            'consolidation_interval': 5000,
        },
        BrainState.DROWSY: {
            'batch_size': 50,
            'temperature': 0.5,
            'exploration_rate': 0.3,
            'consolidation_interval': 10000,
        },
        BrainState.RELAXED: {
            'batch_size': 200,
            'temperature': 1.2,
            'exploration_rate': 0.7,
            'consolidation_interval': 30000,
        },
        BrainState.FOCUSED: {
            'batch_size': 150,
            'temperature': 0.7,
            'exploration_rate': 0.4,
            'consolidation_interval': 60000,
        },
        BrainState.PEAK: {
            'batch_size': 100,
            'temperature': 0.5,
            'exploration_rate': 0.3,
            'consolidation_interval': 45000,
        },
        BrainState.HYPERFOCUS: {
            'batch_size': 100,
            'temperature': 0.5,
            'exploration_rate': 0.3,
            'consolidation_interval': 45000,
        },
    }
    return params.get(state, params[BrainState.FOCUSED])


def run_oscillator_cycle(
    phi: float,
    kappa: float,
    basin_drift: float,
    iterations_since_consolidation: int = 0,
    near_misses_recent: int = 0
) -> Dict:
    """
    Run a full oscillator cycle.

    Returns recommended brain state and search parameters.
    """
    recommended = recommend_brain_state(
        phi, kappa, basin_drift,
        iterations_since_consolidation, near_misses_recent
    )

    neural_oscillators.set_state(recommended)
    search_params = apply_brain_state_to_search(recommended)

    return {
        'recommended_state': recommended.value,
        'current_state': neural_oscillators.to_dict(),
        'search_parameters': search_params,
    }


if __name__ == '__main__':
    print("=" * 60)
    print("NEURAL OSCILLATORS TEST")
    print("=" * 60)

    # Test state transitions
    print("\nTest 1: Default focused state")
    result = neural_oscillators.to_dict()
    print(f"State: {result['current_state']}, κ: {result['kappa']}")

    print("\nTest 2: Transition to relaxed (exploration)")
    neural_oscillators.auto_select_state(SearchPhase.EXPLORATION)
    result = neural_oscillators.to_dict()
    print(f"State: {result['current_state']}, κ: {result['kappa']}")

    print("\nTest 3: Transition to peak (near-miss found)")
    result = run_oscillator_cycle(
        phi=0.75, kappa=64.0, basin_drift=0.1,
        iterations_since_consolidation=10, near_misses_recent=2
    )
    print(f"Recommended: {result['recommended_state']}")
    print(f"Search params: {result['search_parameters']}")

    print("\nTest 4: Need consolidation")
    result = run_oscillator_cycle(
        phi=0.6, kappa=60.0, basin_drift=0.35,
        iterations_since_consolidation=60, near_misses_recent=0
    )
    print(f"Recommended: {result['recommended_state']}")

    print("\n" + "=" * 60)
    print("NEURAL OSCILLATORS READY")
    print("=" * 60)
